client/src/apiUrl.ts

```diff
@@ -0,0 +1,30 @@
+// HackerRank does not run on 'localhost', it creates a vm each time the question runs.
+// The URLs for the web client and api server do not use the standard `:port` in the url, 
+// instead the port is encoded in the hostname.
+//
+// As of 2023-02-29, the port is encoded in the vm's url like so:
+//  https://vm-${uuid-v4}-${port}.us-vmprovider.projects.hrcdn.net
+//
+// This file has a function to get a url to the API from the window location. 
+// It assumes the web's port is 3000 and the server's is 6001.
+// 
+// Per https://support.hackerrank.com/hc/en-us/articles/360000831488-Validating-a-Project-for-Front-End-Back-End-and-Full-Stack-Questions
+// We are using ports 3000 and 6001 to align with their whitelist and to avoid Chrome's list of restricted ports.
+const WEB_PORT = 3000;
+const API_PORT = 6001;
+const WEB_PORT_PATTERN = `-${WEB_PORT}.`; 
+const API_PORT_PATTERN = `-${API_PORT}.`; 
+const API_URL_DEFAULT = `http://localhost:${API_PORT}/api`;
+
+// Set IS_HACKER_RANK to true if running the question within the HackerRank IDE
+const IS_HACKER_RANK = false;
```

Comment:
is there anything in process.env that would indicate if this is being run from within a hackerrank VM?

Reply:
@ogupte Nothing useful in process.env:
```
[SERVER] {
[SERVER]   "DOCKER_MEMORY_LIMIT": "10752M",
[SERVER]   "LANGUAGE": "en_US.UTF-8",
[SERVER]   "npm_config_user_agent": "npm/9.6.4 node/v18.15.0 linux x64 workspaces/false",
[SERVER]   "HOSTNAME": "e7c4a6d6b467",
[SERVER]   "GIT_ASKPASS": "/tmp/vscode-ide/extensions/git/dist/askpass.sh",
[SERVER]   "npm_node_execpath": "/usr/bin/node",
[SERVER]   "SHLVL": "3",
[SERVER]   "BROWSER": "none",
[SERVER]   "npm_config_noproxy": "",
[SERVER]   "HOME": "/home/user",
[SERVER]   "OLDPWD": "/projects/challenge",
[SERVER]   "TERM_PROGRAM_VERSION": "1.70.1",
[SERVER]   "VSCODE_IPC_HOOK_CLI": "/tmp/vscode-ipc-1e9b5d02-995f-463e-ae08-9cddaa19c3a7.sock",
[SERVER]   "npm_package_json": "/projects/challenge/server/package.json",
[SERVER]   "HR_RUNTIME_UID": "1000",
[SERVER]   "VSCODE_GIT_ASKPASS_MAIN": "/tmp/vscode-ide/extensions/git/dist/askpass-main.js",
[SERVER]   "VSCODE_GIT_ASKPASS_NODE": "/tmp/vscode-ide/node",
[SERVER]   "npm_config_userconfig": "/home/user/.npmrc",
[SERVER]   "npm_config_local_prefix": "/projects/challenge/server",
[SERVER]   "COLORTERM": "truecolor",
[SERVER]   "COLOR": "0",
[SERVER]   "npm_config_metrics_registry": "https://registry.npmjs.org/",
[SERVER]   "FORCE_COLOR": "3",
[SERVER]   "_": "/usr/bin/npm",
[SERVER]   "npm_config_prefix": "/projects/challenge/server",
[SERVER]   "TERM": "xterm-256color",
[SERVER]   "npm_config_cache": "/home/user/.npm",
[SERVER]   "npm_config_node_gyp": "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js",
[SERVER]   "PATH": "/projects/challenge/server/node_modules/.bin:/projects/challenge/server/node_modules/.bin:/projects/challenge/node_modules/.bin:/projects/node_modules/.bin:/node_modules/.bin:/usr/lib/node_modules/npm/node_modules/@npmcli/run-script/lib/node-gyp-bin:/projects/challenge/node_modules/.bin:/projects/node_modules/.bin:/node_modules/.bin:/usr/lib/node_modules/npm/node_modules/@npmcli/run-script/lib/node-gyp-bin:/projects/challenge/node_modules/.bin:/projects/node_modules/.bin:/node_modules/.bin:/usr/lib/node_modules/npm/node_modules/@npmcli/run-script/lib/node-gyp-bin:/tmp/vscode-ide/bin/remote-cli:/home/user/.local/bin:/home/user/bin:/home/user/.yarn/bin:/home/user/bin:/home/user/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/projects/challenge/.bin",
[SERVER]   "NODE": "/usr/bin/node",
[SERVER]   "npm_package_name": "nested-list-app-server",
[SERVER]   "HR_RUNTIME_USER": "user",
[SERVER]   "LANG": "en_US.UTF-8",
[SERVER]   "LS_COLORS": "rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.webp=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:",
[SERVER]   "VSCODE_GIT_IPC_HANDLE": "/tmp/vscode-git-753bc1a445.sock",
[SERVER]   "TERM_PROGRAM": "vscode",
[SERVER]   "npm_lifecycle_script": "nodemon ./src/server.ts",
[SERVER]   "DEBIAN_FRONTEND": "noninteractive",
[SERVER]   "npm_package_version": "0.1.0",
[SERVER]   "npm_lifecycle_event": "start",
[SERVER]   "WDS_SOCKET_PORT": "0",
[SERVER]   "VSCODE_GIT_ASKPASS_EXTRA_ARGS": "",
[SERVER]   "npm_config_globalconfig": "/usr/etc/npmrc",
[SERVER]   "npm_config_init_module": "/home/user/.npm-init.js",
[SERVER]   "PWD": "/projects/challenge/server",
[SERVER]   "HR_RUNTIME_GID": "1000",
[SERVER]   "LC_ALL": "en_US.UTF-8",
[SERVER]   "npm_execpath": "/usr/lib/node_modules/npm/bin/npm-cli.js",
[SERVER]   "npm_config_global_prefix": "/projects/challenge/server",
[SERVER]   "npm_command": "run-script",
[SERVER]   "CHROME_BIN": "/usr/bin/google-chrome",
[SERVER]   "VARIANT": "node",
[SERVER]   "INIT_CWD": "/projects/challenge",
[SERVER]   "EDITOR": "vi"
[SERVER] }
```

Reply:
Looks like we might be able to check `process.env.HR_RUNTIME_USER`, that way we can make it easier on the interviewer from having to set this manually at the start of each interview.
```suggestion
const IS_HACKER_RANK = !!process.env.HR_RUNTIME_USER;
```

Reply:
Well, the copy in HackerRank can be edited. I was thinking we would simply change that boolean if we every re-import the code.

But I would prefer a more automatic solution, I'll see what it will entail to push this ENV value through to the front end.